apis/stats_api.py

```diff
@@ -1,360 +1,33 @@
-    results = {}
-
-    db = get_db()
-    with db.cursor(cursor=DictCursor) as cursor:
-        for _, query in queries.items():
-            cursor.execute(query)
-            results.update(cursor.fetchall()[0])
-        return jsonify(results)
-
+def _calc_stats(lock):
+    with lock: 
+        stats.calculate()
 
+@stats_api.get("/calculate")
 @check_admin
-@stats_api.get("/daily")
-def get_daily_stats():
-    queries = {}
-    queries['daily_new_users'] = '''
-    WITH data AS (
-        SELECT 
-            DATE(join_date) AS day,
-            COUNT(*) AS daily_users 
-        FROM users  
-        GROUP BY day 
-    )
-
-    SELECT
-        day,
-        daily_users,
-        SUM(daily_users) OVER (ORDER BY day) AS total 
-    FROM data
-    '''
-
-    queries['daily_sprints'] = '''
-    WITH data AS (
-        SELECT 
-            DATE(start_time) AS day,
-            COUNT(*) AS daily_plays 
-        FROM sprint_runs
-        WHERE start_time IS NOT NULL
-        GROUP BY day 
-    )
-
-    SELECT
-        day,
-        daily_plays,
-        SUM(daily_plays) OVER (ORDER BY day) AS total 
-    FROM data
-    '''
-
-    queries['daily_finished_sprints'] = '''
-    WITH data AS (
-        SELECT 
-            DATE(start_time) AS day,
-            COUNT(*) AS daily_plays 
-        FROM sprint_runs
-        WHERE finished
-        GROUP BY day 
-    )
-
-    SELECT
-        day,
-        daily_plays,
-        SUM(daily_plays) OVER (ORDER BY day) AS total 
-    FROM data
-    '''
-    
-    queries['daily_lobby_runs'] = '''
-    WITH data AS (
-        SELECT 
-            DATE(start_time) AS day,
-            COUNT(*) AS daily_plays 
-        FROM lobby_runs
-        WHERE start_time IS NOT NULL
-        GROUP BY day 
-    )
-
-    SELECT
-        day,
-        daily_plays,
-        SUM(daily_plays) OVER (ORDER BY day) AS total 
-    FROM data
-    '''
-        
-    queries['daily_finished_lobby_runs'] = '''
-    WITH data AS (
-        SELECT 
-            DATE(start_time) AS day,
-            COUNT(*) AS daily_plays 
-        FROM lobby_runs
-        WHERE finished
-        GROUP BY day 
-    )
-
-    SELECT
-        day,
-        daily_plays,
-        SUM(daily_plays) OVER (ORDER BY day) AS total 
-    FROM data
-    '''
-    
-    queries['daily_quick_runs'] = '''
-    WITH data AS (
-        SELECT 
-            DATE(start_time) AS day,
-            COUNT(*) AS daily_plays 
-        FROM quick_runs
-        WHERE start_time IS NOT NULL
-        GROUP BY day 
-    )
-
-    SELECT
-        day,
-        daily_plays,
-        SUM(daily_plays) OVER (ORDER BY day) AS total 
-    FROM data
-    '''
-
-    queries['daily_finished_quick_runs'] = '''
-    WITH data AS (
-        SELECT 
-            DATE(start_time) AS day,
-            COUNT(*) AS daily_plays 
-        FROM quick_runs
-        WHERE finished
-        GROUP BY day 
-    )
+def calculate_stats():
+    if locked(stats.calc_stat_lock):
```

Comment:
Hmmm, due to python GIL I don't think this would actually be hit. For this specific scenario too, it's alright if we have two identical entries. I can add in a non-blocking call in `_calc_stats`, but I'm hesitant to acquire the lock and pass it off. In case that process dies or some exception is hit we don't have a good way to release the lock.

Reply:
I wonder then if a lock is really needed? Could we just have something like "check if we ran in the past minute" or something? Maybe insert a row in the DB with a blank stats to represent a pending job? If it's a whole other process instead of a thread idk if locks are the best fit.

Reply:
Yeah I didn't have to use a lock, I just thought it would be fun to implement like this. It's slightly more robust than doing a check and relatively straightforward as well. Based on these [docs](https://docs.python.org/3/library/multiprocessing.html#synchronization-between-processes) and in my testing it does achieve process-level synchronization. And the reason it's chucked into a new process is to bypass the GIL, same as what celery does.